sw/device/lib/dif/register.h

```diff
@@ -0,0 +1,129 @@
+ */
+inline reg32_t reg32_from_addr(uintptr_t address) {
+  return (reg32_t){
+      .inner_ptr = (volatile uint32_t *)address,
+  };
+}
+
+/**
+ * Reads an aligned word from the MMIO register |base| at the given byte offset.
+ *
+ * This function is guaranteed to commit a read to memory, and will not be
+ * reordered with respect to other register manipulations.
+ *
+ * @param base the register to read from.
+ * @param offset the offset to read at, in bytes.
+ * @return the read word.
+ */
+inline uint32_t reg32_read(reg32_t base, ptrdiff_t offset) {
+  return base.inner_ptr[offset / 4];
+}
+
+/**
+ * Writes an aligned word to the MMIO register |base| at the given byte offset.
+ *
+ * This function is guaranteed to commit a write to memory, and will not be
+ * reordered with respect to other register manipulations.
+ *
+ * @param base the register to write to.
+ * @param offset the offset to write at, in bytes.
+ * @param value the word to write.
+ */
+inline void reg32_write(reg32_t base, ptrdiff_t offset, uint32_t value) {
+  base.inner_ptr[offset / 4] = value;
+}
+
+/**
+ * Clears the bits in |mask| from the MMIO register |base| at the given offset.
+ *
+ * This function has the same guarantees as |reg32_read()| and |reg32_write()|.
+ *
+ * @param base the register to mask.
+ * @param offset the offset to apply the mask at, in bytes.
+ * @param mask the mask to clear from the selected register.
+ */
+inline void reg32_clear_mask(reg32_t base, ptrdiff_t offset, uint32_t mask) {
+  uint32_t value = reg32_read(base, offset);
+  value &= ~mask;
+  reg32_write(base, offset, value);
+}
+
+/**
+ * Sets the bits in |mask| from the MMIO register |base| at the given offset.
+ *
+ * This function has the same guarantees as |reg32_read()| and |reg32_write()|.
+ *
+ * @param base the register to mask.
+ * @param offset the offset to apply the mask at, in bytes.
+ * @param mask the mask to set on the selected register.
+ */
+inline void reg32_set_mask(reg32_t base, ptrdiff_t offset, uint32_t mask) {
+  uint32_t value = reg32_read(base, offset);
+  value |= mask;
+  reg32_write(base, offset, value);
+}
+
+/**
+ * Checks whether the |bit_index|th bit is set in the MMIO register |base| at
+ * the given offset.
+ *
+ * This function has the same guarantees as |reg32_read()| and |reg32_write()|.
+ *
+ * @param base the register to mask.
+ * @param offset the offset to apply the mask at.
+ * @param bit_index the bit to check.
+ */
+inline bool reg32_get_bit(reg32_t base, ptrdiff_t offset, uint32_t bit_index) {
+  return (reg32_read(base, offset) >> bit_index) & 0x1;
+}
+
+/**
+ * Clears the |bit_index|th bit in the MMIO register |base| at the given offset.
+ *
+ * This function has the same guarantees as |reg32_read()| and |reg32_write()|.
+ *
+ * @param base the register to mask.
+ * @param offset the offset to apply the mask at.
+ * @param bit_index the bit to clear.
+ */
+inline void reg32_clear_bit(reg32_t base, ptrdiff_t offset,
```

Comment:
I have incorporated register.{h,c} in my UART DIF - changes all write and read from registers to corresponding functions in register.{h,c}, but linker sees reg32_clear_bit and reg32_set_bit as undefined references. All the other reg32_* routines seem to work (read, write, get).

undefined reference to `reg32_clear_bit'

Reply:
Would you try removing the `inline` specifiers in the `.c` file? I suspect that might be causing issues.

Reply:
I have removed inline specifiers in the c file, but it didn't seem to make a difference. Linker only seems to have problems with `reg32_clear_bit` and `reg32_set_bit`.

Reply:
@mcy my bad, I forgot to include registers.c in meson.build file, and so it seems that compiler was deciding not to inline some of the functions - hence undefined reference. The rest it was inlining and that's why there wasn't any issue.